build: compile 4 webpackRendererConfigs instead of N

georgexu99 opened the pull request:
Co-authored-by: Samuel Attard <marshallofsound@electronjs.org>
Co-authored-by: Keeley Hammond <vertedinde@electronjs.org>

related issue: 
https://github.com/electron/forge/issues/3087

`getRendererConfig` currently returns a separate config per entryPoint: this build speed regression was introduced in https://github.com/electron/forge/commit/1f45e2ca701ec117df932f38e7fe808bdd832e74 when each `entryPoint` gained the capability to set its own `nodeIntegration`

This change routes each entry point into one of 4 buckets such that only 4 configs are generated.

georgexu99 commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -20,6 +20,13 @@ export type ConfigurationFactory = (
+  Web = 'web',
+  ElectronRenderer = 'electron-renderer',
+  ElectronPreload = 'electron-preload',
+  SandboxedPreload = 'web',
```
is this correct

georgexu99 commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -190,7 +197,7 @@ export default class WebpackConfigGenerator {
   }
 
-  async getRendererConfig(entryPoints: WebpackPluginEntryPoint[]): Promise<Configuration[]> {
+  async getRendererConfig2(entryPoints: WebpackPluginEntryPoint[]): Promise<Configuration[]> {
```
will remove this soon

georgexu99 commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -273,4 +280,121 @@ export default class WebpackConfigGenerator {
+      for (const entryPoint of entryPoints) {
+        if (isLocalWindow(entryPoint)) {
+          plugins.push(
+            new HtmlWebpackPlugin({
```
is there a way we can avoid adding a plugin per entrypoint?

MarshallOfSound replied:
No this is just how webpack works

MarshallOfSound commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -20,6 +20,27 @@ export type ConfigurationFactory = (
+      return 'electron-preload';
+    case RendererTarget.ElectronRenderer:
+      return 'electron-renderer';
+    default:
+      return '';
```
The `default` case here should not need to exist, if typescript insists it should exist instead make it `throw`

MarshallOfSound commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -163,114 +180,122 @@ export default class WebpackConfigGenerator {
+    const entryPointsForTarget = {
+      web: [] as (WebpackPluginEntryPointLocalWindow | WebpackPluginEntryPoint)[],
+      electronRenderer: [] as (WebpackPluginEntryPointLocalWindow | WebpackPluginEntryPoint)[],
+      electronPreload: [] as (WebpackPluginEntryPointPreloadOnly | WebpackPluginEntryPointLocalWindow)[],
+      sandboxedPreload: [] as (WebpackPluginEntryPointPreloadOnly | WebpackPluginEntryPointLocalWindow)[],
```
This is incorrect, we should type these as only `WebpackPluginEntryPointPreloadOnly[]`.  If typescript disagrees we should map `entry` into a `preloadOnlyEntry` manually on lines 197 and 206

georgexu99 replied:
can you rephrase what you mean here

MarshallOfSound commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -163,114 +180,122 @@ export default class WebpackConfigGenerator {
+        }
+      }
     }
+    const rendererConfigPromises = [
+      this.buildRendererConfig(entryPointsForTarget.web, RendererTarget.Web),
+      this.buildRendererConfig(entryPointsForTarget.electronRenderer, RendererTarget.ElectronRenderer),
+      this.buildRendererConfig(entryPointsForTarget.electronPreload, RendererTarget.ElectronPreload),
+      this.buildRendererConfig(entryPointsForTarget.sandboxedPreload, RendererTarget.SandboxedPreload),
+    ];
```
```suggestion
    const rendererConfigPromises = await Promise.all([
      this.buildRendererConfig(entryPointsForTarget.web, RendererTarget.Web),
      this.buildRendererConfig(entryPointsForTarget.electronRenderer, RendererTarget.ElectronRenderer),
      this.buildRendererConfig(entryPointsForTarget.electronPreload, RendererTarget.ElectronPreload),
      this.buildRendererConfig(entryPointsForTarget.sandboxedPreload, RendererTarget.SandboxedPreload),
    ]);
```

Let's do the promise awaiting inline, we shouldn't be assigning promises to a value as it potentially let's us experience unhandled rejections with future refactors

MarshallOfSound commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -163,114 +180,122 @@ export default class WebpackConfigGenerator {
+      this.buildRendererConfig(entryPointsForTarget.sandboxedPreload, RendererTarget.SandboxedPreload),
+    ];
+
+    const rendererConfig = await Promise.all(rendererConfigPromises).then((res) => {
```
We shouldn't be using `.then`, we have `async/await`

MarshallOfSound commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -163,114 +180,122 @@ export default class WebpackConfigGenerator {
+    ];
+
+    const rendererConfig = await Promise.all(rendererConfigPromises).then((res) => {
+      return res.filter(function isNotNull<T>(item: T | null): item is T {
```
Extract `isNotNull` to a helper function either at the top level of this file or a separate file

MarshallOfSound commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -163,114 +180,122 @@ export default class WebpackConfigGenerator {
+        __dirname: false,
+        __filename: false,
+      },
+      plugins: [new AssetRelocatorPatch(this.isProd, !!this.pluginConfig.renderer.nodeIntegration)],
```
I know this code existed before but it's semantically incorrect

```suggestion
      plugins: [new AssetRelocatorPatch(this.isProd, target === RendererTarget.ElectronRenderer || target === RendererTarget.ElectronPreload)],
```

MarshallOfSound commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -163,114 +180,122 @@ export default class WebpackConfigGenerator {
+  async buildRendererConfig(entryPoints: WebpackPluginEntryPoint[], target: RendererTarget): Promise<Configuration | null> {
+    const entry: webpack.Entry = {};
+
+    if (entryPoints.length === 0) {
+      return null;
+    }
```
This should be above `const entry`

MarshallOfSound commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -163,114 +180,122 @@ export default class WebpackConfigGenerator {
-        mode: this.mode,
+    if (target === RendererTarget.Web || target === RendererTarget.ElectronRenderer) {
+      for (const entryPoint of entryPoints) {
+        entry[entryPoint.name] = (entryPoint.prefixedEntries || []).concat([isPreloadOnly(entryPoint) ? entryPoint.preload.js : entryPoint.js]);
```
Why is `preload` in here? Neither the `Web` nor `ElectronRenderer` targets should be considering preload values

georgexu99 replied:
TS gets angry and complains that `entryPoint.js` doesn't exist on `WebpackPluginEntryPointPreloadOnly`, even though that should not have made it thrtough the if statement, but TS doesn't know that

MarshallOfSound commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -163,114 +180,122 @@ export default class WebpackConfigGenerator {
+      return webpackMerge(baseConfig, { entry, output, plugins }, rendererConfig || {});
+    } else if (target === RendererTarget.ElectronPreload || target === RendererTarget.SandboxedPreload) {
+      for (const entryPoint of entryPoints) {
+        entry[entryPoint.name] = (entryPoint.prefixedEntries || []).concat([isPreloadOnly(entryPoint) ? entryPoint.preload.js : entryPoint.js]);
```
Why is `.js` here, neither `ElectronPreload` nor `SandboxedPreload` should be referencing anything **other** than the preload JS file

MarshallOfSound commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -163,114 +180,122 @@ export default class WebpackConfigGenerator {
         output: {
           path: path.resolve(this.webpackDir, 'renderer'),
-          filename: '[name]/index.js',
+          filename: 'preload.js',
```
Shouldn't this be `[name]/preload.js`

georgexu99 replied:
wait actually I think it's supposed to be just 'preload.js'
<img width="668" alt="image" src="https://user-images.githubusercontent.com/33054982/205187269-51a15a41-8d4b-402f-a73e-0404ffe2f539.png">

georgexu99 replied:
changed back for now

MarshallOfSound commented on packages/plugin/webpack/src/WebpackPlugin.ts:
```diff
@@ -282,33 +281,30 @@ the generated files). Instead, it is ${JSON.stringify(pj.main)}`);
-        }
-      }
-    }
+    // for (const entryPoint of this.config.renderer.entryPoints) {
+    //   if ((isLocalWindow(entryPoint) && !!entryPoint.preload) || isPreloadOnly(entryPoint)) {
+    //     const stats = await this.runWebpack(
+    //       // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    //       [await this.configGenerator.getPreloadConfigForEntryPoint(entryPoint)]
+    //     );
+
+    //     if (stats?.hasErrors()) {
+    //       throw new Error(`Compilation errors in the preload (${entryPoint.name}): ${stats.toString()}`);
+    //     }
+    //   }
+    // }
```
We can remove these now

MarshallOfSound commented on packages/plugin/webpack/src/WebpackPlugin.ts:
```diff
@@ -321,34 +317,34 @@ the generated files). Instead, it is ${JSON.stringify(pj.main)}`);
-        });
-      }
-    }
+    // for (const entryPoint of this.config.renderer.entryPoints) {
+    //   if ((isLocalWindow(entryPoint) && !!entryPoint.preload) || isPreloadOnly(entryPoint)) {
+    //     const config = await this.configGenerator.getPreloadConfigForEntryPoint(entryPoint);
+    //     config.infrastructureLogging = {
+    //       level: 'none',
+    //     };
+    //     config.stats = 'none';
+    //     await new Promise((resolve, reject) => {
+    //       const tab = logger.createTab(`${entryPoint.name} - Preload`);
+    //       const [onceResolve, onceReject] = once(resolve, reject);
+
+    //       this.watchers.push(
+    //         webpack(config).watch({}, (err, stats) => {
+    //           if (stats) {
+    //             tab.log(
+    //               stats.toString({
+    //                 colors: true,
+    //               })
+    //             );
+    //           }
+
+    //           if (err) return onceReject(err);
+    //           return onceResolve(undefined);
+    //         })
+    //       );
+    //     });
```
Same here can remove

georgexu99 commented:
@MarshallOfSound 
https://github.com/electron/forge/pull/3101/files/9830b8646cbfe857d9649ed2b673b6a04ca6e94d#r1036759743
addressed all comments minus this one, wasn't too sure what you meant. lines 198, 207 are angry- did you want to create a new target? 


lines 250, 277, created a new guard function to assert type let me know if that was not the move

georgexu99 commented on packages/plugin/webpack/package.json:
```diff
@@ -8,7 +8,7 @@
   "typings": "dist/WebpackPlugin.d.ts",
   "scripts": {
-    "test": "xvfb-maybe mocha --config ../../../.mocharc.js test/**/*_spec.ts"
+    "test": "xvfb-maybe mocha --config ../../../.mocharc.js test/**/*_spec.ts test/*_spec.ts"
```
old tests under the package root weren't being run, such as webpackConfig_spec.ts

georgexu99 commented on packages/plugin/webpack/src/util/rendererTypeUtils.ts:
```diff
@@ -36,7 +40,7 @@ export const isNoWindowEntries = (entries: WebpackPluginEntryPoint[]): entries i
 export const isPreloadOnlyEntries = (entries: WebpackPluginEntryPoint[]): entries is WebpackPluginEntryPointPreloadOnly[] => {
   for (const entry of entries) {
-    if (!isPreloadOnly(entry)) {
+    if (!hasPreloadScript(entry)) {
```
changed from previous iteration's `isPreloadOnly` since now, an entry can have `html` + `preload.js` in this bucket

georgexu99 commented on packages/plugin/webpack/test/WebpackConfig_spec.ts:
```diff
@@ -483,7 +331,7 @@ describe('WebpackConfigGenerator', () => {
       const generator = new WebpackConfigGenerator(config, mockProjectDir, false, 3000);
       const webpackConfig = await generator.getRendererConfig(config.renderer.entryPoints);
-      expect(webpackConfig[0].target).to.equal('electron-preload');
+      expect(webpackConfig[0].target).to.equal('web');
```
this test uses an entry with `nodeintegration: false` (default) , so witih the new changes it gets put into `sandboxedpreload` bucket which resolves to `web` target rather than `electron-preload` (unsandboxed preloads)

georgexu99 commented on packages/plugin/webpack/test/WebpackConfig_spec.ts:
```diff
@@ -535,7 +383,7 @@ describe('WebpackConfigGenerator', () => {
       const generator = new WebpackConfigGenerator(config, mockProjectDir, true, 3000);
       const webpackConfig = await generator.getRendererConfig(config.renderer.entryPoints);
-      expect(webpackConfig[0].target).to.deep.equal('electron-preload');
+      expect(webpackConfig[0].target).to.deep.equal('web');
```
see above

georgexu99 commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -163,114 +181,115 @@ export default class WebpackConfigGenerator {
+      sandboxedPreload: [] as WebpackPluginEntryPointPreloadOnly[],
+    };
 
-    const rendererConfig = await this.resolveConfig(entryPoint.preload.config || this.pluginConfig.renderer.config);
```
@MarshallOfSound do we ever need the individual entryPoint's preload config? We no longer handle this

georgexu99 commented:
Currently potentially having issue w/ asset relocator patch for the failing tests :( path/to/x is not being shortened to _dirname

MarshallOfSound commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -20,6 +20,28 @@ export type ConfigurationFactory = (
+  SandboxedPreload,
+}
+function isNotNull<T>(item: T | null): item is T {
+  return !!item;
```
```suggestion
  return item !== null;
```

MarshallOfSound commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -20,6 +20,28 @@ export type ConfigurationFactory = (
+  return !!item;
+}
+
+function rendererTargetToWebpackTarget(target: RendererTarget): string {
```
Is there a better type than `string` for a webpack target name?

MarshallOfSound commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -163,114 +181,151 @@ export default class WebpackConfigGenerator {
+        }
+      }
     }
+    const rendererConfigs = await Promise.all(
+      (
+        await this.buildRendererConfigs(entryPointsForTarget.web, RendererTarget.Web)
+      ).concat(
+        await this.buildRendererConfigs(entryPointsForTarget.electronRenderer, RendererTarget.ElectronRenderer),
+        await this.buildRendererConfigs(entryPointsForTarget.electronPreload, RendererTarget.ElectronPreload),
+        await this.buildRendererConfigs(entryPointsForTarget.sandboxedPreload, RendererTarget.SandboxedPreload)
+      )
+    );
```
```suggestion
    const rendererConfigs = (await Promise.all(
      [
        await this.buildRendererConfigs(entryPointsForTarget.web, RendererTarget.Web),
        await this.buildRendererConfigs(entryPointsForTarget.electronRenderer, RendererTarget.ElectronRenderer),
        await this.buildRendererConfigs(entryPointsForTarget.electronPreload, RendererTarget.ElectronPreload),
        await this.buildRendererConfigs(entryPointsForTarget.sandboxedPreload, RendererTarget.SandboxedPreload)
      ]
    )).reduce((configs, allConfigs) => allConfigs.concat(configs));
```

georgexu99 replied:
ah very clean

MarshallOfSound commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -163,114 +181,151 @@ export default class WebpackConfigGenerator {
+      plugins:
+        target === RendererTarget.ElectronPreload
+          ? []
+          : [new webpack.ExternalsPlugin('commonjs2', ['electron', 'electron/renderer', 'electron/common', 'events', 'timers', 'url'])],
```
While we're here can these modules be extracted to a constant somewhere nice

MarshallOfSound commented on packages/plugin/webpack/src/util/rendererTypeUtils.ts:
```diff
@@ -15,3 +15,27 @@ export const isPreloadOnly = (entry: WebpackPluginEntryPoint): entry is WebpackP
 };
+
+export const hasPreloadScript = (entry: WebpackPluginEntryPoint): entry is WebpackPluginEntryPointPreloadOnly => {
+  return !!(entry as any).preload;
```
```suggestion
  return 'preload' in entry;
```

georgexu99 replied:
ah very clean

MarshallOfSound commented on packages/plugin/webpack/test/AssetRelocatorPatch_spec.ts:
```diff
@@ -145,14 +145,13 @@ describe('AssetRelocatorPatch', () => {
+        outDir: rendererOut,
         jsPath: path.join(rendererOut, 'main_window/preload.js'),
-        nativeModulesString: '__webpack_require__.ab = __dirname + "/native_modules/"',
+        nativeModulesString: `__webpack_require__.ab = ${JSON.stringify(rendererOut)} + "/native_modules/"`,
```
This seems wrong, doesn't that mean we have a hardcoded path (from the build machine) being injected into the final output 🤔

georgexu99 replied:
If I am understanding this code correctly, we do in fact have a hardcoded path being injected in XD
<img width="711" alt="image" src="https://user-images.githubusercontent.com/33054982/219211357-748e129b-34fa-4035-89be-4eeaae157214.png">

MarshallOfSound commented on packages/plugin/webpack/test/AssetRelocatorPatch_spec.ts:
```diff
@@ -199,13 +198,13 @@ describe('AssetRelocatorPatch', () => {
+        outDir: rendererOut,
         jsPath: path.join(rendererOut, 'main_window/preload.js'),
-        nativeModulesString: '.ab=__dirname+"/native_modules/"',
+        nativeModulesString: '.ab=require("path").resolve(require("path").dirname(__filename),"..")+"/native_modules/"',
```
Did you figure out what generates this path? This could probably be `resolve(__dirname, '..')` instead?

georgexu99 replied:
For this test, we generate the path in `injectedProductionDirnameCode()` within `AssetRelocatorPatch.ts`

MarshallOfSound commented on packages/plugin/webpack/test/WebpackConfig_spec.ts:
```diff
@@ -27,51 +27,6 @@ const sampleWebpackConfig = {
 };
 
 describe('WebpackConfigGenerator', () => {
-  describe('rendererTarget', () => {
```
Is there a replacement for these tests, we still decide targets based on entrypoint objects right?

MarshallOfSound commented on packages/plugin/webpack/test/WebpackConfig_spec.ts:
```diff
@@ -308,113 +263,6 @@ describe('WebpackConfigGenerator', () => {
     });
   });
 
-  describe('getPreloadConfigForEntryPoint', () => {
```
Same here, we have new functions for generating configs, we should test those new functions if we're deleting the old tests

kyleshay commented:
Gentle bump @MarshallOfSound — I tried building forge from this commit and was going to test it against my steps from this comment: https://github.com/electron/forge/issues/3087#issuecomment-1331187076 to validate the change, but was running into an issue with linking, I think.

georgexu99 commented on packages/plugin/webpack/src/WebpackConfig.ts:
```diff
@@ -275,7 +275,7 @@ export default class WebpackConfigGenerator {
       }
     }
-    return webpackMerge(baseConfig, rendererConfig || {}, { entry, output, plugins });
+    return webpackMerge({ entry, output, plugins }, rendererConfig || {}, baseConfig);
```
Webpack Merge prioritizes right -> left:

we determine the config target within the baseConfig (i.e 'web', 'electron-preload') and not the one generated from `rendererConfig = this.pluginConfig.renderer.config`.

georgexu99 replied:
edit: maybe I was wrong and shouldn't have touched this

georgexu99 replied:
https://yarnpkg.com/package?name=webpack-merge#:~:text=webpack%2Dmerge%20provides%20a%20merge,values%20within%20a%20function%20again.

within the first code example, a comment mentions that keys to the right take precedence. 
I think these need to be moved because otherwise, the separate targets that this PR evaluates do not get translated into the final webpack config (ex. prior to this change, `target: electron-preload` target in the baseconfig gets overwritten  by `target: web` in the rendererConfig, even if its a preload-only entrypoint

I'm gonna dig a bit deeper and test whether I can make a functioning webpack app with these changes

VerteDinde reviewed (approved):
Approved on the condition that we document or revert the webpack merge change above - excited to have this in!
